File: database/database-impl/src/main/java/io/ehdev/conrad/database/api/internal/ReverseApiCommitComparator.java
package io.ehdev.conrad.database.api.internal;

import io.ehdev.conrad.database.model.project.commit.ApiCommitModel;

import java.util.Comparator;

class ReverseApiCommitComparator implements Comparator<ApiCommitModel> {

    @Override
    public int compare(ApiCommitModel o1, ApiCommitModel o2) {
        if (o1 == o2) {
            return 0;
        }

        String[] split1 = o1.getVersion().split("\\.");
        String[] split2 = o2.getVersion().split("\\.");
        int sharedParts = Math.min(split1.length, split2.length);
        for (int i = 0; i < sharedParts; i++) {
            if (split1[i].compareTo(split2[i]) != 0) {
                return -1 * split1[i].compareTo(split2[i]);
            }
        }

        if (split1.length > split2.length) {
            return -1;
        } else if (split1.length < split2.length) {
            return 1;
        } else {
            return 0;
        }
    }
}

File: database/database-api-model/src/main/kotlin/io/ehdev/conrad/database/model/comparator/ReverseApiCommitComparator.kt
package io.ehdev.conrad.database.model.comparator

import io.ehdev.conrad.database.model.project.commit.ApiCommitModel
import java.util.*
import kotlin.collections.dropLastWhile
import kotlin.collections.toTypedArray
import kotlin.text.isEmpty
import kotlin.text.split
import kotlin.text.toRegex

class ReverseApiCommitComparator : Comparator<ApiCommitModel> {

    override fun compare(o1: ApiCommitModel, o2: ApiCommitModel): Int {
        if (o1 == o2) {
            return 0
        }

        val split1 = o1.version!!.split("\\.".toRegex()).dropLastWhile({ it.isEmpty() }).toTypedArray()
        val split2 = o2.version!!.split("\\.".toRegex()).dropLastWhile({ it.isEmpty() }).toTypedArray()
        val sharedParts = Math.min(split1.size, split2.size)
        for (i in 0..sharedParts - 1) {
            if (split1[i].compareTo(split2[i]) != 0) {
                return -1 * split1[i].compareTo(split2[i])
            }
        }

        if (split1.size > split2.size) {
            return -1
        } else if (split1.size < split2.size) {
            return 1
        } else {
            return 0
        }
    }
}
